packages/devtools_app/lib/src/shared/ui/search.dart

```diff
@@ -62,6 +61,14 @@ mixin SearchControllerMixin<T extends DataSearchStateMixin> {
   /// Duration.zero (default) disables debounce
   Duration? get debounceDelay => null;
 
+  /// Text field controller for the [SearchField] that this instance of
+  /// [SearchControllerMixin] controls.
+  SearchTextEditingController? searchTextFieldController;
+
+  /// Focus node for the [SearchField] that this instance of
+  /// [SearchControllerMixin] controls.
+  FocusNode? searchFieldFocusNode;
+
```

Comment:
When would these be `null` (why are they nullable?)

Reply:
These are nulled out in dispose and re-initialized in initSearch. The enclosing controller mixin will live longer than the search field for which the text field and the focus node are used. However, there are several places in the controller that need to access the text field controller and the focus node, so we store them in the controller, and then init and dispose them with the widget lifecycle (this is what SearchFieldMixin does). I don't love this solution, but it seemed cleaner than what we had.

Reply:
Got it, that makes sense. I'm mostly trying to figure out if there is a way having all the null-assert `!s` for these fields. Would it be possible to change these to be `late` fields (initialized in `initSearch`) and disposed of but not set to `null` during dispose? Or do they need to be reset back to `null`?

Reply:
Setting to null ensures that we don't try to dispose an already disposed object, whereas using late wouldn't protect against that. For example, we dispose these objects in initSearch before re-initializing them, and we wouldn't be able to do that if we used late.

Reply:
Got it, thanks! Optional - switch these (and `rawKeyboardFocusNode` below) to private nullable fields, and add corresponding non-nullable getter for them. Besides that, LGTM!